// lwc-services/lib/commands/serve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const cli_ux_1 = require("cli-ux");
const fs = require("fs");
const path = require("path");
const webpack = require("webpack");
const webpackMerge = require("webpack-merge");
const lwcConfig_1 = require("../config/lwcConfig");
const webpack_config_1 = require("../config/webpack.config");
const serve_1 = require("../messages/serve");
const logger_1 = require("../utils/logger");
class Serve extends command_1.Command {
    async run() {
        const { flags } = this.parse(Serve);
        // tslint:disable-next-line: no-console
        console.clear();
        logger_1.welcome();
        const BUILD_DIR = flags.directory ? flags.directory : lwcConfig_1.lwcConfig.buildDir;
        lwcConfig_1.lwcConfig.server.host = flags.host ? flags.host : lwcConfig_1.lwcConfig.server.host;
        lwcConfig_1.lwcConfig.server.port = flags.port ? flags.port : lwcConfig_1.lwcConfig.server.port;
        // Override for SaaS deployments
        if (process.env.PORT) {
            lwcConfig_1.lwcConfig.server.port = Number(process.env.PORT);
        }
        // Check if given source directory exists. If not we're exiting.
        if (!fs.existsSync(BUILD_DIR)) {
            logger_1.log(serve_1.messages.errors.no_build_dir, BUILD_DIR);
            return;
        }
        // Check if custom webpack config is passed, and if it really exists.
        if (flags.webpack) {
            if (!fs.existsSync(flags.webpack)) {
                logger_1.log(serve_1.messages.errors.no_webpack);
                return;
            }
        }
        let webpackConfig = webpack_config_1.generateWebpackConfig('production');
        lwcConfig_1.lwcConfig.devServer.contentBase = lwcConfig_1.lwcConfig.sourceDir;
        webpackConfig.devServer = Object.assign({}, lwcConfig_1.lwcConfig.devServer, lwcConfig_1.lwcConfig.server);
        // Merging custom webpack config file
        if (flags.webpack) {
            logger_1.log(serve_1.messages.logs.custom_configuration);
            const webpackConfigCustom = require(path.resolve(process.cwd(), flags.webpack));
            webpackConfig = webpackMerge.smart(webpackConfig, webpackConfigCustom);
        }
        if (flags.host && flags.host !== lwcConfig_1.lwcConfig.server.host) {
            webpackConfig.devServer.host = flags.host;
        }
        if (flags.port && flags.port !== lwcConfig_1.lwcConfig.server.port) {
            webpackConfig.devServer.port = flags.port;
        }
        // Lazy loading
        const WebpackDevServer = require('webpack-dev-server');
        // tslint:disable-next-line: no-unused
        const compiler = webpack(webpackConfig);
        const app = new WebpackDevServer(compiler, webpackConfig.devServer);
        app.listen(webpackConfig.devServer.port, webpackConfig.devServer.host, () => {
            const protocol = 'http';
            const url = `${protocol}://${webpackConfig.devServer.host}:${webpackConfig.devServer.port}`;
            logger_1.log(serve_1.messages.logs.local_server_listening, url);
            if (flags.open) {
                // tslint:disable-next-line: no-floating-promises
                cli_ux_1.default.open(url);
            }
        });
    }
}
Serve.description = serve_1.messages.description;
Serve.examples = serve_1.messages.help.examples;
Serve.flags = {
    help: command_1.flags.help({ char: 'h' }),
    directory: command_1.flags.string({
        char: 'd',
        description: serve_1.messages.flags.directory,
        default: lwcConfig_1.lwcConfig.buildDir
    }),
    host: command_1.flags.string({
        char: 'i',
        description: serve_1.messages.flags.host,
        default: lwcConfig_1.lwcConfig.server.host
    }),
    open: command_1.flags.boolean({
        char: 'o',
        description: serve_1.messages.flags.open,
        default: lwcConfig_1.lwcConfig.server.open
    }),
    port: command_1.flags.integer({
        char: 'p',
        description: serve_1.messages.flags.port,
        default: lwcConfig_1.lwcConfig.server.port
    }),
    webpack: command_1.flags.string({
        char: 'w',
        description: serve_1.messages.flags.webpack
    })
};
exports.default = Serve;
